### PR TITLE
Finite Strain class for 2D AxisymmetricRZ

### LINKED ISSUE #5142: Migrate Axisymmetric material from Solid Mechanics to Tensor Mechanics
The start of migrating materials + associated kernels from Solid Mechanics to Tensor Mechanics.  This issue focuses on the axisymmetricRZ "strain calculator" in Solid Mechanics with plans to incorporate the capability through the ComputeStrainBase material in Tensor Mechanics.  Under discussion/ advisement with @jasondhales @bwspenc 

Assign to me

### PR DESCRIPTION
**Pull request by sapitts**
Addresses and closes #5142 with addition of the Finite Strain version of the Axisymmetric 'strain calculator' migrated from Solid Mechanics.  A 3D model of a 1/8 sphere is included for comparison with the 2D RZ simulation to more completely mirror the work done in Solid Mechanics; the 3D model is marked as 'heavy.'  This finite strain RZ test exodiff passes with the Solid Mechanics nonlinearRZ test.

This commit also includes a test for the axisymmetric Rank-4 C_ijlk fill method and some whitespace error corrections.

### CODE REVIEW

**Comment by sapitts** [comment]
Need to update with the new constructors; will push the corrected code later today.

File: modules/tensor_mechanics/include/materials/Compute2DFiniteStrain.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * Compute2DFiniteStrain defines a strain increment and a rotation increment
+ * for finite strains in 2D geometries, handling the out of plane strains.
+ * Compute2DFiniteStrain contains a virtual method to define the strain_zz
+ * as a nonzero value in the inherited classes ComputePlaneFiniteStrain
+ * ComputeGeneralizedPlaneFiniteStrain and ComputeAxisymmetricRZFiniteStrain.
+ */
+class Compute2DFiniteStrain : public ComputeFiniteStrain
+{
+public:
+  Compute2DFiniteStrain(const std:: string & name, InputParameters parameters);
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/Compute2DFiniteStrain.h]: You should probably update this to use the new constructor:
```
Compute2DFiniteStrain(const InputParameters & parameters);
```

Comment (dschwen) [inline modules/tensor_mechanics/include/materials/Compute2DFiniteStrain.h]: This applies to all constructors below as well.

File: modules/tensor_mechanics/src/materials/Compute2DFiniteStrain.C [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+Compute2DFiniteStrain::computeProperties()
+{
+  //Method from Rashid, 1993
+  std::vector<RankTwoTensor> Fhat;
+  Fhat.resize(_qrule->n_points());
+  RankTwoTensor ave_Fhat;
+  Real volume(0);
+  Real ave_dfgrd_det;
+
+  ave_dfgrd_det=0.0;
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/Compute2DFiniteStrain.C]: ... :-)

**Comment by dschwen** [comment]
Ok, looks good to me, and I didn;t see your comment about the new constructors. I think this can be merges once the constructors are updated.

**Comment by jwpeterson** [comment]
I'll take a crack at getting this up-to-date with the latest constructors so that it can be merged...

**Comment by sapitts** [comment]
Thanks John, I am sorry about my delay in getting the needed corrections in.

**Comment by jwpeterson** [comment]
That's no problem, we are helping out with the constructor updates wherever we can.

**Comment by dschwen** [comment]
Ugh, I reviewed this and missed the fact that c159bee adds stateful properties to Tensor Mechanics even in the small strain case! This is bad.

**Comment by sapitts** [comment]
The disp_x value is needed to calculate the stress_{\theta} in the
axisymmetric code, and in the finite strain code also the old value is
needed; unfortunately I wasn't able to set the old value in the inherited
ComputeFiniteStrain class while setting the current value in the base
ComputeStrainBase class.  I think Pritam ran into this issue of declaring
old properties farther down the inheritance chain this summer as well?
Looking forward to seeing your input parameter magic, or I could pull in
solely the displacement in the ComputeSmallStrain class and the disp and
disp_old values in the ComputeFiniteStrain class.

--

Stephanie Pitts
Graduate Student, Mechanical Engineering, and Instructor
Computational Mechanics & Materials Science Laboratory
Washington State University, sapitts@wsu.edu

**Comment by dschwen** [comment]
@sapitts #5643 implements the magic.

**Comment by dschwen** [comment]
@sapitts I really don't know why the stateful disp_x is not declared in the derived class. I'll try changing that.

**Comment by sapitts** [comment]
@dschwen probably because I got a bit carried away with finally understanding how the inherited classes worked--I also vaguely remember being concerned about correctly implementing the vector of displacements.  Certainly declaring the disp_x value in the AxisymmetricSmallStrain would clean up ComputeSmallStrain while keeping the disp_x dependent needed for the Axisymmetric case.  I'm a bit more concerned about the AxisymmetricFiniteStrain case, but it should be a similar fix

**Comment by sapitts** [comment]
@dschwen One cavet:  if Derek wants to implement the rotation about the X axis, we'll need the disp[1] value to compute stress_zz, so perhaps it makes sense to add a comment somewhere if you pull in only disp[0] instead of the full "vector"?

**Pull request merged**